src/main/java/net/fabricmc/loader/impl/discovery/ModPrioSorter.java

```diff
@@ -0,0 +1,245 @@
+
+				// quick check if it might match
+				if (!potentiallyOverlappingIds.contains(cmpId)
+						&& (cmpMod.getProvides().isEmpty() || Collections.disjoint(potentiallyOverlappingIds, cmpMod.getProvides()))) {
+					continue;
+				}
+
+				int cmp = compareOverlappingIds(mod, cmpMod, Integer.MAX_VALUE);
+
+				if (cmp < 0) { // mod needs to be after cmpMod, move mod forward
+					//System.out.printf("found candidate for %d at %d (before %s)%n", i, j, cmpMod);
+					earliestIdx = j;
+				} else if (cmp != Integer.MAX_VALUE) { // cmpMod has at least the same prio, don't search past it
+					break;
+				}
+			}
+
+			if (earliestIdx >= 0) {
+				//System.out.printf("move %d to %d (before %s)%n", i, earliestIdx, mods.get(earliestIdx));
+				mods.remove(i);
+				mods.add(earliestIdx, mod);
+			}
+
+			potentiallyOverlappingIds.clear();
+		}
+	}
+
+	private static final Comparator<ModCandidate> comparator = new Comparator<ModCandidate>() {
+		@Override
+		public int compare(ModCandidate a, ModCandidate b) {
+			return ModPrioSorter.compare(a, b);
+		}
+	};
+
+	private static int compare(ModCandidate a, ModCandidate b) {
+		// descending sort prio (less/earlier is higher prio):
+		// root mods first, lower id first, higher version first, less nesting first, parent cmp
+
+		if (a.isRoot()) {
+			if (!b.isRoot()) {
+				return -1; // only a is root
+			}
+		} else if (b.isRoot()) {
+			return 1; // only b is root
+		}
+
+		// sort id asc
+
+		int idCmp = a.getId().compareTo(b.getId());
+		if (idCmp != 0) return idCmp;
+
+		// sort version desc (lower version later)
+		int versionCmp = b.getVersion().compareTo(a.getVersion());
+		if (versionCmp != 0) return versionCmp;
+
+		// sort nestLevel asc
+		int nestCmp = a.getMinNestLevel() - b.getMinNestLevel(); // >0 if nest(a) > nest(b)
+		if (nestCmp != 0) return nestCmp;
+
+		if (a.isRoot()) return 0; // both root
+
+		// find highest priority parent, if it is not shared by both a+b the one that has it is deemed higher prio
+		return compareParents(a, b);
+	}
+
+	private static int compareParents(ModCandidate a, ModCandidate b) {
+		assert !a.getParentMods().isEmpty() && !b.getParentMods().isEmpty();
+
+		ModCandidate minParent = null;
+
+		for (ModCandidate mod : a.getParentMods()) {
+			if (minParent == null || mod != minParent && compare(minParent, mod) > 0) {
+				minParent = mod;
+			}
+		}
+
+		assert minParent != null;
+		boolean found = false;
+
+		for (ModCandidate mod : b.getParentMods()) {
+			if (mod == minParent) { // both a and b have minParent
+				found = true;
+			} else if (compare(minParent, mod) > 0) { // b has a higher prio parent than a
+				return 1;
+			}
+		}
+
+		return found ? 0 : -1; // only a has minParent if !found, so only a has the highest prio parent
+	}
+
+	private static int compareOverlappingIds(ModCandidate a, ModCandidate b, int noMatchResult) {
```

Comment:
Returning a `@Nullable Integer` might be better here.

Reply:
OptionalInt?